boot/freeldr/freeldr/custom.c

```diff
@@ -320,58 +432,97 @@ VOID OptionMenuCustomBootLinux(VOID)
 {
     ULONG_PTR SectionId;
     CHAR SectionName[100];
     CHAR BootDriveString[20];
     CHAR BootPartitionString[20];
     CHAR ReactOSSystemPath[200];
     CHAR ReactOSARCPath[200];
-    CHAR ReactOSOptions[200];
+    PCHAR ReactOSOptions = MmAllocateMemoryWithType(sizeof(CHAR) * 200, LoaderSystemCode);
     TIMEINFO* TimeInfo;
     OperatingSystemItem OperatingSystem;
 
-    RtlZeroMemory(SectionName, sizeof(SectionName));
-    RtlZeroMemory(BootDriveString, sizeof(BootDriveString));
-    RtlZeroMemory(BootPartitionString, sizeof(BootPartitionString));
-    RtlZeroMemory(ReactOSSystemPath, sizeof(ReactOSSystemPath));
-    RtlZeroMemory(ReactOSOptions, sizeof(ReactOSOptions));
+    if (!IniSectionName)
+    {
+        RtlZeroMemory(SectionName, sizeof(SectionName));
+        RtlZeroMemory(BootDriveString, sizeof(BootDriveString));
+        RtlZeroMemory(BootPartitionString, sizeof(BootPartitionString));
+        RtlZeroMemory(ReactOSSystemPath, sizeof(ReactOSSystemPath));
+        RtlZeroMemory(ReactOSOptions, sizeof(ReactOSOptions));
+    }
+    else
+    {
+        CHAR BootType[80] = ANSI_NULL;
+
+        /* Try to open the operating system section in the .ini file */
+        if (IniOpenSection(IniSectionName, &SectionId))
+        {
+            /* Try to read the boot type */
+            IniReadSettingByName(SectionId, "BootType", BootType, sizeof(BootType));
+        }
+        /* If no or invalid type specified, we should not be there; fail */
+        if ((BootType[0] == ANSI_NULL) ||
+            (_stricmp(BootType, "Windows") != 0) ||
+            (_stricmp(BootType, "WindowsNT40") != 0) ||
+            (_stricmp(BootType, "Windows2003") != 0))
+        {
+            return;
+        }
 
-    if (!UiEditBox(BootDrivePrompt, BootDriveString, 20))
-        return;
+        /* Load the settings */
+        // TODO? Maybe use DissectArcPath(CHAR *ArcPath, CHAR *BootPath, UCHAR* BootDrive, ULONG* BootPartition) to get back to the small elements.
+        IniReadSettingByName(SectionId, "SystemPath", ReactOSARCPath, sizeof(ReactOSARCPath));
+        IniReadSettingByName(SectionId, "Options", ReactOSOptions, sizeof(ReactOSOptions));
+    }
 
-    if (!UiEditBox(BootPartitionPrompt, BootPartitionString, 20))
-        return;
+    if (!IsEditCmdLine)
+    {
+        if (!UiEditBox(BootDrivePrompt, BootDriveString, sizeof(BootDriveString)))
+            return NULL;
 
-    if (!UiEditBox(ReactOSSystemPathPrompt, ReactOSSystemPath, 200))
-        return;
+        if (!UiEditBox(BootPartitionPrompt, BootPartitionString, sizeof(BootPartitionString)))
+            return NULL;
 
-    if (!UiEditBox(ReactOSOptionsPrompt, ReactOSOptions, 200))
-        return;
+        if (!UiEditBox(ReactOSSystemPathPrompt, ReactOSSystemPath, sizeof(ReactOSSystemPath)))
+            return NULL;
+    }
+
+    if (!IsSetup)
+    {
+        if (UiEditBox(ReactOSOptionsPrompt, ReactOSOptions, sizeof(ReactOSOptions)))
```

Comment:
`sizeof` here is now `sizeof(PVOID)` since `ReactOSOptions` is no longer an array.

Reply:
Good catch!